src/deroff.rs

```diff
@@ -1216,18 +1222,68 @@ fn test_do_tbl() {
     assert!(deroffer.do_tbl());
     assert_eq!(deroffer.tblTab, "a");
     assert_eq!(deroffer.s, ";");
+    assert_eq!(deroffer.tblstate, TblState::Format);
 
     // </Options>
 
+    // <Format>
+
     let mut deroffer = Deroffer::new();
     deroffer.tblstate = TblState::Format;
+    deroffer.s = "Hello World!".into();
+    assert!(deroffer.do_tbl());
+    assert_ne!(deroffer.tblstate, TblState::Data);
+    assert_eq!(deroffer.output.take(), "\n");
+    assert!(deroffer.s.is_empty());
 
+    let mut deroffer = Deroffer::new();
+    deroffer.tblstate = TblState::Format;
     deroffer.s = "".into();
+    assert!(deroffer.do_tbl());
+    assert_ne!(deroffer.tblstate, TblState::Data);
+    assert_eq!(deroffer.output.take(), "\n");
+    assert!(deroffer.s.is_empty());
+
+    let mut deroffer = Deroffer::new();
+    deroffer.tblstate = TblState::Format;
+    deroffer.s = "Hello World!.foo bar!".into();
+    assert!(deroffer.do_tbl());
+    assert_eq!(deroffer.tblstate, TblState::Data);
+    assert_eq!(deroffer.s, ".foo bar!");
+    assert_eq!(deroffer.output.take(), "\n");
+
+    let mut deroffer = Deroffer::new();
+    deroffer.tblstate = TblState::Format;
+    deroffer.s = "\n".into();
+    assert!(deroffer.do_tbl());
+    assert_ne!(deroffer.tblstate, TblState::Data);
+    assert_eq!(deroffer.s, "\n");
+    assert_eq!(deroffer.output.take(), "\n");
+
+    let mut deroffer = Deroffer::new();
+    deroffer.tblstate = TblState::Format;
+    deroffer.s = ".".into();
+    assert!(deroffer.do_tbl());
+    assert_eq!(deroffer.tblstate, TblState::Data);
+    assert_eq!(deroffer.s, ".");
+    assert_eq!(deroffer.output.take(), "\n");
+
+    // </Format>
+    // <Data>
 
     let mut deroffer = Deroffer::new();
     deroffer.tblstate = TblState::Data;
+    deroffer.tblTab = "a".into();
 
-    deroffer.s = "".into();
+    deroffer.s = "HelloaWorld!".into();
+    assert!(deroffer.do_tbl());
+    // This also calls text, so this checks that works correctly
+    // "Hello\tWorld!" passed to self.text
+    assert_eq!(deroffer.output.take(), "Hello\tWorld!");
+
+    // </Data>
+
+    // TODO: Add tests for the rest of the functions
```

Comment:
Isn't this finished?

Reply:
This test is finished, but there are a bunch of functions without tests (text being one of them)

Reply:
I think we can leave that to a separate patch. This should be good enough to cover `do_tbl`.